tools/hdf_dump_translation_dataset.py

```diff
@@ -0,0 +1,325 @@
+#!/usr/bin/env python3
+
+import sys
+import argparse
+import gzip
+import pickle
+
+import numpy
+import h5py
+from operator import itemgetter
+from itertools import islice, zip_longest
+
+UNKNOWN_LABEL = "<UNK>"
+POSTFIX = " </S>"
+
+
+class HDFTranslationDatasetCreator(object):
+  """
+  Creates the same HDF file as hdf_dump.py called on a TranslationDataset, but does so much faster
+  and using much less memory.
+
+  Input parameters are basically the files in a TranslationDataset folder. Additionally, the number of lines has to
+  be given.
+  """
+
+  def __init__(self, hdf_file, source_file, target_file, source_vocabulary, target_vocabulary, number_of_lines,
+               compression_method=None, line_buffer_size=100000, data_buffer_size=5000000):
+    """
+    :param str hdf_file: filename for the hdf file being created
+    :param str source_file: filename of the source text file
+    :param str target_file: filename of the target text file
+    :param str source_vocabulary: filename of the source vocabulary (in pickle format)
+    :param str target_vocabulary: filename of the target vocabulary (in pickle format)
+    :param int number_of_lines: line count in source_file and target_file
+    :param str compression_method: Optional compression method as supported by h5py.File.create_dataset().
+        Applied to the main data only ('inputs' and 'target/data/classes').
+    :param int line_buffer_size: number of corpus lines to read and process at once
+    :param int data_buffer_size: space to reserve inside the hdf file at once, in numbers of integers
+    """
+
+    self.hdf_file_name = hdf_file
+    self.hdf_file = None
+    self.compression_method = compression_method
+
+    self.source_file_handle = self._open_file(source_file)
+    self.target_file_handle = self._open_file(target_file)
+
+    source_vocabulary = self._read_vocabulary(source_vocabulary)
+    target_vocabulary = self._read_vocabulary(target_vocabulary)
+
+    self._vocabularies = {"source": source_vocabulary, "target": target_vocabulary}
+    self._vocabulary_sizes = {"source": len(source_vocabulary), "target": len(target_vocabulary)}
+    self._unknown_ids = {"source": source_vocabulary[UNKNOWN_LABEL], "target": target_vocabulary[UNKNOWN_LABEL]}
+
+    self.number_of_lines = number_of_lines
+    self.line_buffer_size = line_buffer_size
+    self.data_buffer_size = data_buffer_size
+
+    self._number_of_processed_lines = 0
+    self._write_offsets = {"source": 0, "target": 0}
+
+  def __del__(self):
+    self.source_file_handle.close()
+    self.target_file_handle.close()
```

Comment:
Normally you never need (nor should) define `__del__`. Also in this case, I don't see the point. If the GC catches the object, it will also catch these handles, and then also close them. Do you really need that? Depending on the Python interpreter, it anyway has no effect.